server/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java

```diff
@@ -129,15 +130,25 @@ protected void metadataToXContent(XContentBuilder builder, Params params) throws
     }
 
     @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+    protected XContentBuilder toXContent(XContentBuilder builder, Params params, Set<Throwable> seenExceptions) throws IOException {
         Throwable ex = ExceptionsHelper.unwrapCause(this);
         if (ex != this) {
-            generateThrowableXContent(builder, params, this);
+            generateThrowableXContent(builder, params, this, seenExceptions);
         } else {
             // We don't have a cause when all shards failed, but we do have shards failures so we can "guess" a cause
             // (see {@link #getCause()}). Here, we use super.getCause() because we don't want the guessed exception to
             // be rendered twice (one in the "cause" field, one in "failed_shards")
-            innerToXContent(builder, params, this, getExceptionName(), getMessage(), getHeaders(), getMetadata(), super.getCause());
+            innerToXContent(
```

Comment:
Just throwing out an idea here, an alternative could be to move tracking of observed instances into the builder itself.  E.g. using something as simple (and manual) as a method `checkAndRecord`: `object -> boolean` 

```java
if(builder.checkAndRecord(throwable) == false){
   builder.field(TRUNCATED, true);
   return;
}
```

Besides limiting the interfaces affected by the change, it has the benefit that it would cover `failed_shards` as well and potentially remove the need for using `super.getCause()` here: the 2nd appearance would simply be truncated without any further ado.

Though, not sure if that justifies for that and if it would be useful anywhere else beyond exception graphs.

Reply:
I was surprised that I couldn't find any existing patterns for cycle/duplicate detection in our XContent rendering (I'm sure we must have _some_, but I didn't work out where it was)

Having something built into the builder would be helpful, but I was hesitant to change core interfaces without having a conversation first.

Reply:
We don't have anything general purpose for xcontent. The assumption is that we are not serializing things outside of our control, so we shouldn't have cycles/SOF problems. when we do have an object that is out of our control (eg returned by a script), we use `CollectionUtils.checkNoSelfReferences`, which effectively does this same pattern of capturing seen references. If we intend to keep with reference checking, perhaps that method could be moved/adjusted to be more general purpose so as not to duplicate it, really we just need a way to iterate/getref.

With that said, ref checking does have a heavier weight runtime cost than depth counting. If we want a general purpose guard for xcontent, depth counting sounds more appealing. However, IMO the complexity as noted by Tim of when and how to limit depth is not worthwhile. As noted above, we control all objects that are serialized. This specific case is one of self reference, it is a bug in exception generation, so let's treat it like that, and have self reference checking (I'm not concerned with the runtime cost since it is already an exception, and in general the exception chain should be relatively short and doesn't fan like a collection).